README.md
NECモバイルバックエンド基盤: Node.js 用 SSE Push SDK
====================================================

NECモバイルバックエンド基盤(BaaS)の Node.js 向け SSE Push
クライアントライブラリです。

本ライブラリは Node.js 専用です。ブラウザでは利用できません。

準備
----

本ライブラリを使用するアプリケーションディレクトリ(package.jsonが存在する
ディレクトリ)に移動し、本ライブラリを npm install でインストールしてください。

    $ npm install @nec-baas/ssepush-node --save

使用方法
--------

サンプルが example/client.js にありますので、詳細はこちらを確認ください。

### インスタンスの生成

最初に、本ライブラリを require してください。
(BaaS ライブラリの require と初期化も必要です)

    const SsePush = require('@nec-baas/ssepush-node').SsePush;

以下手順で SsePush インスタンスを生成します。

    const ssePush = new SsePush(Nebula);
    // デバイストークン設定
    ssePush.setDeviceToken(deviceToken);

デバイストークンは、デバイスに固有の識別子です。UUIDなどを使用することを
推奨します(SsePush.uuid()で生成可能)。デバイストークンは同一デバイス
では原則変更しないように、適宜保存するなどしてください。

### 受信チャネル設定

受信チャネルを設定する場合は、setChannels で指定します。

    ssePush.setChannels(['channel1', 'channel2'])

### ハートビート間隔・監視設定

ハートビート間隔の設定は setHeartbeatInterval で行います。
値はサーバ側の設定に合わせてください。

    ssePush.setHeartbeatInterval(30);

ハートビート間隔の2倍の時間が経過してもハートビート受信ができない場合は、
onError でエラー通知されます。

なお、ハートビート間隔を設定しない場合は、ハートビート監視は行われません。

### 登録・受信開始

BaaSサーバへのインスタレーション登録と Push 待ち受けを開始します。
ユーザがログイン状態になっている場合は、インスタレーションは自動的にユーザ
に紐づけされます。

    // BaaSサーバへの登録と Push 待ち受けを開始
    ssePush.start({
    	onMessage: (message) => {
    		// 受信処理
    	},
    	onError: (error) => {
    		// エラー処理
    	}
    });

SSE Push メッセージを受信するたびに onMessage が呼び出されます。
引数でメッセージが渡されます。メッセージは以下のような JSON object
で、本文は "data" に格納されます。

    {
        "type": "message",
        "data": "This is a test.",
        "lastEventId": "",
        "origin": "https://baas.example.com"
    }
 
エラー時は onError が呼び出されますので、適宜回復処理を行ってください。
ステータスコードは error.status に格納されます。

オプション
---------

Proxy サーバを使用する場合は、SsePush コンストラクタの第二引数にオプション
を指定してください (注: 認証付きの Proxy は使用できません)

    const options = { proxy: 'http://proxy.example.com:8080' }
    const ssePush = new SsePush(Nebula, options);

HTTPS接続でかつサーバ側が自己署名証明書を使用している場合は、以下のオプションを指定します。
(非推奨)

    const options = { rejectUnauthorized: false }

その他のオプションについては、[EventSource](https://github.com/EventSource/eventsource)
の README を参照してください。

OSSライセンス
-------------

本モジュールは EventSource v1.0.5 を改造したものを含んでいます。
(lib/eventsource.js)
https://github.com/EventSource/eventsource

改造内容は以下のとおりです。

* oncomment コールバックを追加
* options に agent を追加

ライセンス条件は LICENSE.eventsource を参照してください。


ChangeLog.txt
2018/10/01 : v7.5.0

        * js-sdk v7.5.0の適用

2018/08/27 : v7.5.0-rc.1

        * js-sdk v7.5.0-rc.1の適用

2018/06/20 : v7.0.3

        * 機能追加・改善
          - 送信者許可設定API追加
          - Push受信待機完了通知追加

2018/05/16 : v7.0.2

        * 機能追加・改善
          - CONNECTメソッドの Host ヘッダ値修正
            - tunnel モジュールを tunnel-fork に変更
          - js-sdk v7.0.2の適用

2018/03/29 : v7.0.0

        * js-sdk v7.0.1の適用

2018/03/09 : v7.0.0 rc1

        * 機能追加・改善
          - [8591] パッケージ名を @nec-baas/ssepush-node に変更

2017/12/13 : v7.0.0-beta1

        * 不具合修正
          - [7701] サーバ証明書検証が行われない問題を修正 (eventsource.js 側の不具合回避)

2017/10/04

        - Proxy 設定変更。HTTPS接続時は自動的に tunnel agent を使用する。

2017/09/21 : v1.0.1

        - shutdown() 時に heart beat タイマを停止する処理を追加
        - start() 時に古い eventSource を shutdown する処理を追加
        - デバッグ用ログ追加
        
2017/07/14 : v1.0.0

        - 初版


tests/ssepushTest.js
const SsePush = require('../lib/ssepush.js').SsePush;

const chai = require('chai');
const assert = chai.assert;
const expect = chai.expect;

describe("SsePush init", () => {
    const sse = new SsePush({}, {});

    it("初期化", () => {
        expect(sse.channels).have.lengthOf(0);
        expect(sse.allowedSenders).have.lengthOf(0);
        expect(sse.heartbeatInterval).equal(0);
    });

    it("デバイストークン設定", () => {
        expect(sse.setDeviceToken("TOKEN")).equal(sse);
        expect(sse.deviceToken).equal("TOKEN");
    });

    it("チャネル設定", () => {
        expect(sse.setChannels(["a","b"])).equal(sse);
        expect(sse.channels).deep.equal(["a", "b"]);
    });

    it("送信者許可設定", () => {
        expect(sse.setAllowedSenders(["g:group1","g:group2"])).equal(sse);
        expect(sse.allowedSenders).deep.equal(["g:group1","g:group2"]);
    });
    
    it("ハートビート間隔設定", () => {
        expect(sse.setHeartbeatInterval(60)).equal(sse);
        expect(sse.heartbeatInterval).equal(60);
    });
});

describe("_createRegistrationData", () => {
    const sse = new SsePush({}, {});

    it("正常に registration data が生成されること", () => {
        sse.setDeviceToken("TOKEN1");
        sse.setChannels(["c1", "c2"]);
        sse.setAllowedSenders(["g:group1","g:group2"]);
        
        const data = sse._createRegistrationData();

        expect(data._deviceToken).equal("TOKEN1");
        expect(data._channels).deep.equal(["c1", "c2"]);
        expect(data._allowedSenders).deep.equal(["g:group1","g:group2"]);
        expect(data._pushType).equal("sse");
    });
});

describe("_convertProxyOptions", () => {
    const sse = new SsePush({}, {});
    const proxy = "http://proxy.example.com:8080";

    it("Proxy未指定時は何もしないこと", () => {
        const options = {};
        sse._convertProxyOptions("xxx", options);
        expect(options).deep.equal({});
    });

    it("HTTPS接続でない場合は何もしないこと", () => {
        const options = {proxy: proxy};
        sse._convertProxyOptions("http://example.com/sse", options);
        expect(options.proxy).equal(proxy);
        expect(options).not.have.property("agent");
    });

    it("HTTPS接続時に tunnel agent が設定されること", () => {
        const options = {proxy: proxy, rejectUnauthorized: false};
        sse._convertProxyOptions("https://example.com/sse", options);

        expect(options).not.have.property("proxy");
        const agent = options.agent;
        expect(agent).not.be.null;
        expect(agent.options.proxy.host).equal("proxy.example.com");
        expect(agent.options.proxy.port).equal(8080);
        expect(agent.options.rejectUnauthorized).be.false;
        expect(agent.defaultPort).equal(443);
    })
});


example/client.js
const Nebula = require('@nec-baas/jssdk').Nebula;
const SsePush = require('../lib/ssepush.js').SsePush;
//const SsePush = require('baas-ssepush-nodejs').SsePush;

const config = require('./config.js');
const proxy = config.proxy;

// BaaS 初期化
if (proxy != null) {
    Nebula.setHttpProxy(proxy);
    Nebula.setHttpsProxy(proxy);
}
Nebula.initialize(config);

// SSE Push インスタンス生成
const options = {};
if (proxy != null) {
    options.proxy = "http://" + proxy.host + ":" + proxy.port;
}
const ssePush = new SsePush(Nebula, options);

// デバイストークン設定
ssePush.setDeviceToken(config.deviceToken);

// ハートビート間隔設定
ssePush.setHeartbeatInterval(30);

// 登録・受信開始
console.log("Start connect: deviceToken = " + config.deviceToken);
ssePush.start({
    onMessage: (message) => {
        console.log("push message: " + JSON.stringify(message));
    },
    onError: (error) => {
        console.log("ERROR: " + JSON.stringify(error));
    }
});





lib/eventsource.js
var original = require('original')
var parse = require('url').parse
var events = require('events')
var https = require('https')
var http = require('http')
var util = require('util')

var httpsOptions = [
  'pfx', 'key', 'passphrase', 'cert', 'ca', 'ciphers',
  'rejectUnauthorized', 'secureProtocol', 'servername'
]

/**
 * Creates a new EventSource object
 *
 * @param {String} url the URL to which to connect
 * @param {Object} [eventSourceInitDict] extra init params. See README for details.
 * @api public
 **/
function EventSource (url, eventSourceInitDict) {
  var readyState = EventSource.CONNECTING
  Object.defineProperty(this, 'readyState', {
    get: function () {
      return readyState
    }
  })

  Object.defineProperty(this, 'url', {
    get: function () {
      return url
    }
  })

  var self = this
  self.reconnectInterval = 1000

  function onConnectionClosed () {
    if (readyState === EventSource.CLOSED) return
    readyState = EventSource.CONNECTING
    _emit('error', new Event('error'))

    // The url may have been changed by a temporary
    // redirect. If that's the case, revert it now.
    if (reconnectUrl) {
      url = reconnectUrl
      reconnectUrl = null
    }
    setTimeout(function () {
      if (readyState !== EventSource.CONNECTING) {
        return
      }
      connect()
    }, self.reconnectInterval)
  }

  var req
  var lastEventId = ''
  if (eventSourceInitDict && eventSourceInitDict.headers && eventSourceInitDict.headers['Last-Event-ID']) {
    lastEventId = eventSourceInitDict.headers['Last-Event-ID']
    delete eventSourceInitDict.headers['Last-Event-ID']
  }

  var discardTrailingNewline = false
  var data = ''
  var eventName = ''

  var reconnectUrl = null

  function connect () {
    var options = parse(url)
    var isSecure = options.protocol === 'https:'
    options.headers = { 'Cache-Control': 'no-cache', 'Accept': 'text/event-stream' }
    if (lastEventId) options.headers['Last-Event-ID'] = lastEventId
    if (eventSourceInitDict && eventSourceInitDict.headers) {
      for (var i in eventSourceInitDict.headers) {
        var header = eventSourceInitDict.headers[i]
        if (header) {
          options.headers[i] = header
        }
      }
    }

    // Legacy: this should be specified as `eventSourceInitDict.https.rejectUnauthorized`,
    // but for now exists as a backwards-compatibility layer
    options.rejectUnauthorized = !(eventSourceInitDict && !eventSourceInitDict.rejectUnauthorized)

    // If specify http proxy, make the request to sent to the proxy server,
    // and include the original url in path and Host headers
    var useProxy = eventSourceInitDict && eventSourceInitDict.proxy
    if (useProxy) {
      var proxy = parse(eventSourceInitDict.proxy)
      isSecure = proxy.protocol === 'https:'

      options.protocol = isSecure ? 'https:' : 'http:'
      options.path = url
      options.headers.Host = options.host
      options.hostname = proxy.hostname
      options.host = proxy.host
      options.port = proxy.port
    }

    // If https options are specified, merge them into the request options
    if (eventSourceInitDict && eventSourceInitDict.https) {
      for (var optName in eventSourceInitDict.https) {
        if (httpsOptions.indexOf(optName) === -1) {
          continue
        }

        var option = eventSourceInitDict.https[optName]
        if (option !== undefined) {
          options[optName] = option
        }
      }
    }

    // tunnel agent for https with proxy
    if (eventSourceInitDict && eventSourceInitDict.agent) {
      options.agent = eventSourceInitDict.agent;
    }

    // Pass this on to the XHR
    if (eventSourceInitDict && eventSourceInitDict.withCredentials !== undefined) {
      options.withCredentials = eventSourceInitDict.withCredentials
    }

    req = (isSecure ? https : http).request(options, function (res) {
      // Handle HTTP errors
      if (res.statusCode === 500 || res.statusCode === 502 || res.statusCode === 503 || res.statusCode === 504) {
        _emit('error', new Event('error', {status: res.statusCode}))
        onConnectionClosed()
        return
      }

      // Handle HTTP redirects
      if (res.statusCode === 301 || res.statusCode === 307) {
        if (!res.headers.location) {
          // Server sent redirect response without Location header.
          _emit('error', new Event('error', {status: res.statusCode}))
          return
        }
        if (res.statusCode === 307) reconnectUrl = url
        url = res.headers.location
        process.nextTick(connect)
        return
      }

      if (res.statusCode !== 200) {
        _emit('error', new Event('error', {status: res.statusCode}))
        return self.close()
      }

      readyState = EventSource.OPEN
      res.on('close', function () {
        res.removeAllListeners('close')
        res.removeAllListeners('end')
        onConnectionClosed()
      })

      res.on('end', function () {
        res.removeAllListeners('close')
        res.removeAllListeners('end')
        onConnectionClosed()
      })
      _emit('open', new Event('open'))

      // text/event-stream parser adapted from webkit's
      // Source/WebCore/page/EventSource.cpp
      var buf = ''
      res.on('data', function (chunk) {
        buf += chunk

        var pos = 0
        var length = buf.length

        while (pos < length) {
          if (discardTrailingNewline) {
            if (buf[pos] === '\n') {
              ++pos
            }
            discardTrailingNewline = false
          }

          var lineLength = -1
          var fieldLength = -1
          var c

          for (var i = pos; lineLength < 0 && i < length; ++i) {
            c = buf[i]
            if (c === ':') {
              if (fieldLength < 0) {
                fieldLength = i - pos
              }
            } else if (c === '\r') {
              discardTrailingNewline = true
              lineLength = i - pos
            } else if (c === '\n') {
              lineLength = i - pos
            }
          }

          if (lineLength < 0) {
            break
          }

          parseEventStreamLine(buf, pos, fieldLength, lineLength)

          pos += lineLength + 1
        }

        if (pos === length) {
          buf = ''
        } else if (pos > 0) {
          buf = buf.slice(pos)
        }
      })
    })

    req.on('error', onConnectionClosed)
    if (req.setNoDelay) req.setNoDelay(true)
    req.end()
  }

  connect()

  function _emit () {
    if (self.listeners(arguments[0]).length > 0) {
      self.emit.apply(self, arguments)
    }
  }

  this._close = function () {
    if (readyState === EventSource.CLOSED) return
    readyState = EventSource.CLOSED
    if (req.abort) req.abort()
    if (req.xhr && req.xhr.abort) req.xhr.abort()
  }

  function parseEventStreamLine (buf, pos, fieldLength, lineLength) {
    if (lineLength === 0) {
      if (data.length > 0) {
        var type = eventName || 'message'
        _emit('message', new MessageEvent(type, { // ### always emit 'message', not type.
          data: data.slice(0, -1), // remove trailing newline
          lastEventId: lastEventId,
          origin: original(url)
        }))
        data = ''
      }
      eventName = void 0
    } else if (fieldLength > 0) {
      var noValue = fieldLength < 0
      var step = 0
      var field = buf.slice(pos, pos + (noValue ? lineLength : fieldLength))

      if (noValue) {
        step = lineLength
      } else if (buf[pos + fieldLength + 1] !== ' ') {
        step = fieldLength + 1
      } else {
        step = fieldLength + 2
      }
      pos += step

      var valueLength = lineLength - step
      var value = buf.slice(pos, pos + valueLength)

      if (field === 'data') {
        data += value + '\n'
      } else if (field === 'event') {
        eventName = value
      } else if (field === 'id') {
        lastEventId = value
      } else if (field === 'retry') {
        var retry = parseInt(value, 10)
        if (!Number.isNaN(retry)) {
          self.reconnectInterval = retry
        }
      }
    } else if (fieldLength === 0) { // ### handling comment
      _emit('comment', null);
    }
  }
}

module.exports = EventSource

util.inherits(EventSource, events.EventEmitter)
EventSource.prototype.constructor = EventSource; // make stacktraces readable

['open', 'error', 'message', 'comment'].forEach(function (method) { // ### add oncomment
  Object.defineProperty(EventSource.prototype, 'on' + method, {
    /**
     * Returns the current listener
     *
     * @return {Mixed} the set function or undefined
     * @api private
     */
    get: function get () {
      var listener = this.listeners(method)[0]
      return listener ? (listener._listener ? listener._listener : listener) : undefined
    },

    /**
     * Start listening for events
     *
     * @param {Function} listener the listener
     * @return {Mixed} the set function or undefined
     * @api private
     */
    set: function set (listener) {
      this.removeAllListeners(method)
      this.addEventListener(method, listener)
    }
  })
})

/**
 * Ready states
 */
Object.defineProperty(EventSource, 'CONNECTING', {enumerable: true, value: 0})
Object.defineProperty(EventSource, 'OPEN', {enumerable: true, value: 1})
Object.defineProperty(EventSource, 'CLOSED', {enumerable: true, value: 2})

EventSource.prototype.CONNECTING = 0
EventSource.prototype.OPEN = 1
EventSource.prototype.CLOSED = 2

/**
 * Closes the connection, if one is made, and sets the readyState attribute to 2 (closed)
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/API/EventSource/close
 * @api public
 */
EventSource.prototype.close = function () {
  this._close()
}

/**
 * Emulates the W3C Browser based WebSocket interface using addEventListener.
 *
 * @param {String} type A string representing the event type to listen out for
 * @param {Function} listener callback
 * @see https://developer.mozilla.org/en/DOM/element.addEventListener
 * @see http://dev.w3.org/html5/websockets/#the-websocket-interface
 * @api public
 */
EventSource.prototype.addEventListener = function addEventListener (type, listener) {
  if (typeof listener === 'function') {
    // store a reference so we can return the original function again
    listener._listener = listener
    this.on(type, listener)
  }
}

/**
 * Emulates the W3C Browser based WebSocket interface using removeEventListener.
 *
 * @param {String} type A string representing the event type to remove
 * @param {Function} listener callback
 * @see https://developer.mozilla.org/en/DOM/element.removeEventListener
 * @see http://dev.w3.org/html5/websockets/#the-websocket-interface
 * @api public
 */
EventSource.prototype.removeEventListener = function removeEventListener (type, listener) {
  if (typeof listener === 'function') {
    listener._listener = undefined
    this.removeListener(type, listener)
  }
}

/**
 * W3C Event
 *
 * @see http://www.w3.org/TR/DOM-Level-3-Events/#interface-Event
 * @api private
 */
function Event (type, optionalProperties) {
  Object.defineProperty(this, 'type', { writable: false, value: type, enumerable: true })
  if (optionalProperties) {
    for (var f in optionalProperties) {
      if (optionalProperties.hasOwnProperty(f)) {
        Object.defineProperty(this, f, { writable: false, value: optionalProperties[f], enumerable: true })
      }
    }
  }
}

/**
 * W3C MessageEvent
 *
 * @see http://www.w3.org/TR/webmessaging/#event-definitions
 * @api private
 */
function MessageEvent (type, eventInitDict) {
  Object.defineProperty(this, 'type', { writable: false, value: type, enumerable: true })
  for (var f in eventInitDict) {
    if (eventInitDict.hasOwnProperty(f)) {
      Object.defineProperty(this, f, { writable: false, value: eventInitDict[f], enumerable: true })
    }
  }
}


lib/ssepush.ts
/*
 * NEC Mobile Backend Platform : SSE Push(Node.js) library version 7.0.0
 *
 * Copyright (C) 2014-2018, NEC Corporation.
 * All rights reserved.
 */
import Timer = NodeJS.Timer;

const Nebula = require('@nec-baas/jssdk');
const EventSource = require('./eventsource.js');
const url = require('url');

/**
 * SSE Push 受信クラス
 */
export class SsePush {
    private Nebula: any;
    private options: any;
    private channels: string[];
    private allowedSenders: string[];
    private heartbeatInterval: number;

    private deviceToken: string;
    private callback: any;
    private eventSource: any;
    private timer: Timer;

    /**
     * コンストラクタ
     * @param Nebula Nebulaインスタンス
     * @param options オプション。詳細は EventSource のマニュアル参照。
     * @constructor
     */
    constructor(Nebula: any, options: any) {
        SsePush._assertNotNull(Nebula, "Nebula");

        this.Nebula = Nebula;
        this.options = options;

        this.channels = [];
        this.allowedSenders = [];
        this.heartbeatInterval = 0;
    }

    static _assertNotNull(value: any, name: string) {
        if (value == null) {
            throw new Error(name + " must not be null");
        }
    }

    /**
     * デバイストークン（デバイス固有の識別子)をセットする。
     * UUID を使用することを推奨
     * @param deviceToken デバイストークン
     * @returns {SsePush} this
     */
    setDeviceToken(deviceToken: string) {
        SsePush._assertNotNull(deviceToken, "deviceToken");

        this.deviceToken = deviceToken;
        return this;
    }

    /**
     * 購読する channel 名の配列を指定する
     * @param channels channel名の配列
     * @returns {SsePush} this
     */
    setChannels(channels: string[]) {
        SsePush._assertNotNull(channels, "channels");

        this.channels = channels;
        return this;
    }

    /**
     * Push 送信を許可する allowedSender 名の配列を指定する
     * @param allowedSenders allowedSender名の配列
     * @returns {SsePush} this
     */
    setAllowedSenders(allowedSenders: string[]) {
        SsePush._assertNotNull(allowedSenders, "allowedSenders");

        this.allowedSenders = allowedSenders;
        return this;
    }
    
    /**
     * ハートビート間隔(秒)を指定する。
     * <p>
     * ハートビート間隔の2倍時間経過してもハートビートを受信できない場合は、
     * onError でエラー通知する。未設定時はハートビート監視しない。
     * @param interval ハートビート間隔(秒)
     * @returns {SsePush} this
     */
    setHeartbeatInterval(interval: number) {
        this.heartbeatInterval = interval;
        return this;
    }

    /**
     * SSE Push インスタレーション登録と受信を開始する。
     * callback には以下の2つのメソッドを指定すること。
     * <ul>
     *     <li>onMessage(event): メッセージ受信時に呼び出される。
     *     <li>onError(error): エラー時に呼び出される。error には status, statusText, responseText の3フィールドが設定される。
     * </ul>
     * <p>
     * 注) タイミングによっては、401 Unauthorized エラーになる場合がある。
     * (Pushサーバ側に認証情報が登録される前に接続が開始された場合)。
     * この場合はアプリケーション側でリトライする必要がある。
     * @param callback コールバック
     */
    start(callback: any) {
        SsePush._assertNotNull(callback, "callback");

        this.shutdown(); // shutdown old eventSource.

        console.log("SsePush: start");
        this.callback = callback;
        this._startRegister();
    }

    /**
     * 接続をシャットダウンする。
     * ハートビートタイマも停止する。
     */
    shutdown() {
        this._stopHeartbeatTimer();
        if (this.eventSource != null) {
            console.log("SsePush: shutdown: closing eventSource.");
            this.eventSource.close();
            this.eventSource = null;
        } else {
            console.log("SsePush: shutdown: already closed.");
        }
    }

    /**
     * BaaS サーバに対し Installation の登録を行う。
     * 登録が完了したら _startReceiver を呼び出して受信を開始する。
     * @private
     */
    _startRegister() {
        const req = new this.Nebula.HttpRequest(this.Nebula, "/push/installations");

        const data = this._createRegistrationData();

        req.setMethod("POST");
        req.setContentType("application/json");

        req.setData(data);

        req.execute()
            .then((r: any) => {
                const response = JSON.parse(r);
                const sse = response._sse;
                const uri = sse.uri;
                const username = sse.username;
                const password = sse.password;

                console.log("SsePush: Installation registered: uri=" + uri + " username=" + username);

                // TODO: 登録後時間を開けずに接続すると 401 エラーになる場合があるため、
                // ウェイトを入れる。
                setTimeout(() => {
                    this._startReceiver(uri, username, password);
                }, 3000);
            })
            .catch((err: any) => {
                console.error("SsePush: Installation register failed: " + JSON.stringify(err));
                this.callback.onError(err);
            });
    }

    _createRegistrationData() {
        SsePush._assertNotNull(this.deviceToken, "deviceToken");
        SsePush._assertNotNull(this.channels, "channels");
        SsePush._assertNotNull(this.allowedSenders, "allowedSenders");
        
        const data: any = {
            "_osType": "js",
            "_osVersion": "Unknown",
            "_deviceToken": this.deviceToken,
            "_pushType": "sse",
            "_channels": this.channels,
            "_appVersionCode": -1,
            "_appVersionString": "1.0.0",
            "_allowedSenders": this.allowedSenders
        };
        return data;
    }

    /**
     * SSE Push の受信を開始する。
     * @param uri SSE Push サーバ URI
     * @param username ユーザ名
     * @param password パスワード
     * @private
     */
    _startReceiver(uri: string, username: string, password: string) {
        console.log("SsePush: Start receiver");

        const options: any = {
            rejectUnauthorized: true // デフォルト値 (#7701 対応)
        };
        if (this.options != null) {
            Object.assign(options, this.options);
        }
        if (options.headers == null) {
            options.headers = {}
        }
        // Basic認証用のヘッダ指定
        options.headers['Authorization'] = 'Basic ' + new Buffer(username + ':' + password).toString('base64');

        // Proxy 関連設定
        this._convertProxyOptions(uri, options);

        const eventSource = new EventSource(uri, options);
        this.eventSource = eventSource;

        eventSource.onopen = () => {
            console.log("SsePush: Connection established.");
            this.callback.onMessage("SsePush: Connection established.");
            this._restartHeartbeatTimer();
        };
        eventSource.onmessage = this.callback.onMessage;
        eventSource.onerror = this.callback.onError;
        eventSource.oncomment = () => {
            this._onHeartbeat();
        };
    }

    /**
     * Proxy オプションの変換。
     * SSE Push サーバ接続が HTTPS の場合、agent を使用するように変更する。
     * @param uri SSE Push サーバ URI
     * @param options オプション
     * @private
     */
    _convertProxyOptions(uri: string, options: any) {
        if (options.proxy == null) return; // no proxy options

        const parsedUrl = url.parse(uri);
        if (parsedUrl.protocol !== 'https:') return; // not https

        const proxyUrl = url.parse(options.proxy);
        const agentOptions: any = {
            proxy: {
                host: proxyUrl.hostname,
                port: parseInt(proxyUrl.port)
            }
        };
        if (options.rejectUnauthorized != null) {
            agentOptions.rejectUnauthorized = options.rejectUnauthorized;
        }
        const tunnel = require('tunnel-fork');
        const agent = tunnel.httpsOverHttp(agentOptions);
        agent.defaultPort = 443;

        options.agent = agent;
        delete options.proxy;
    }

    /**
     * ハートビート受信処理
     * @private
     */
    _onHeartbeat() {
        //console.log("heartbeat.");
        this._restartHeartbeatTimer()
    }

    /**
     * ハートビートタイマ設定
     * @private
     */
    _restartHeartbeatTimer() {
        this._stopHeartbeatTimer();
        if (this.heartbeatInterval <= 0) return;

        this.timer = setTimeout(() => {
            console.log("SsePush: heartbeat timer timed out, shut down.");
            this.shutdown();
            this.callback.onError({
                'type': 'error',
                'status': 500,
                'message': 'heartbeat timer timed out.'
            });
        }, this.heartbeatInterval * 2 * 1000);
    }

    /**
     * ハートビートタイマ停止
     * @private
     */
    _stopHeartbeatTimer() {
        if (this.timer != null) {
            clearTimeout(this.timer);
        }
    }

    /**
     * ユーティリティ。UUID生成。
     * @returns {string} UUID
     */
    uuid() {
        const S4 = () => {
            return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
        };
        return S4() + S4() + "-" + S4() + "-" + S4() + "-" + S4() + S4() + S4();
    }
}
